### PR TITLE
Telefonica Proposal

### PR DESCRIPTION
**Pull request by monamok**
Telefónica Proposal based on the conversation [here](https://github.com/camaraproject/NumberVerification/pull/2#issuecomment-1326082499)

### CODE REVIEW

**Review by bigludo7 — COMMENTED** [review]
Hello
We have an API close to this proposal in Orange. 2 comments:

1) Could we align the POST body with MC structure:
`{
  "mc_claims": {
    "device_msisdn": "+33640049871"
  }
}`
(and I guess device_hashed_msisdn attribute for the second resource)
It will not change the philosophy API and not add complexity but allow people familiar with this pattern to have it.

2) We have in Orange another resource to retrieve directly the msisdn from the access token (GET operation). This is relevant for some UC - you do not have this requirement on your side ?

Thanks.

**Comment by DT-DawidWroblewski** [comment]
@bigludo7 for a resource call this is something that is common for vMSISDN, so we can align on a POST call to a resource that contains a body you mentioned:

`{ "mc_claims": { "device_msisdn": "+33640049871" } }`

for GET operation to extract MSISDN out of Access Token - this is one of the solutions available on a market. However, this call should take place during the device authentication process initiated with a request coming to /authorize endpoint. If MSISDN extraction is fine, then "code" is returned and we're following OIDC flow.

The reason why it should take place in auth process is thanks to this approach, we're hiding various means of extracting MSISDN out of...whatever (access token, specific token, HE). For the developer, this is handy, because there is a common flow (OpenID Connect) and MNOs hides complexity.

**Comment by monamok** [comment]
Hello,

About the first comment:

>     1. Could we align the POST body with MC structure:
>        `{   "mc_claims": { "device_msisdn": "+33640049871"   } }`
>        (and I guess device_hashed_msisdn attribute for the second resource)
>        It will not change the philosophy API and not add complexity but allow people familiar with this pattern to have it.

The reason why we prefer to separate them in two endpoints is to have control over the scopes and be able to manage the permissions separately. In order to have an intial version 0.X we can have them a single endpoint. 
But I find it a bit confusing naming it "mc_claims" as it's a new API and not MC. I agree that there might be people familiar with the concept but not necesarily. My suggestion is to use a more user-fiendly name.

About the second point:

>     2. We have in Orange another resource to retrieve directly the msisdn from the access token (GET operation). This is relevant for some UC - you do not have this requirement on your side ?


This is related to issue [#101](https://github.com/camaraproject/WorkingGroups/issues/101) that is still under discussion. TEF is strongly against returning PI and sharing that information with 3d-parties. But if already have this UC (I think DT also mentioned that they do as well) we can include it as a separate GET endpoint getting the number from the access token.

Access token can be result of OIDC flow but we insist to have endpoints separeted from the authentication flow.

I will update the yaml and upload it so you can comment on it.

**Comment by monamok** [comment]
I changed the yaml file so there is an single endpoint to verify the number using oneOf structure where it can receive plain phone number or hashed phone number. 
Another option is to reuse the same parameter for clear number as well as hashed number and then have a flag to indicate if it's hashed or not. In that case the request body looks like this:
{
    "phone_number": "xxx"
    "hashed": "false"
}
As you can see in my previous comment, I think mentioning mc_claims in this API is not a good idea because we're trying to agree on having a resource specific API and there is no claims here that we should refer to.

I didn't include the share endpoint (getting the number from the token and returning it) as it's not in our roadmap but if you consider it necesarry you can include it to this version as it's a 0.x version and for the v1.0 we can decide later once issue #101 get resolved.

**Comment by monamok** [comment]
Dear all, could you please review my previous comment and the last changes that were made. Today is the deadline and from Telefónica we hope we can meet it together.

**Comment by monamok** [comment]
Dear all, as it was agreed in this morning meeting I updated the proposal.

@bigludo7 I added a new endpoint for the share operation as you requested. It's a GET operation and returns the phone number. As I already mentioned, sharing PI is not in Telefónica roadmap but having in mind that you have this use case and the commonality issue for the PI is still open and under discussion, we can have it in this version and we'll adjust it later for v1.0, depending on that issue's result. 
Please check it out if it covers your requirements. Thank you.

**Review by bigludo7 — APPROVED** [review]
Thanks @monamok  - This is good for my perspective.

**Comment by monamok** [comment]
> Thanks @monamok - This is good for my perspective.

Thank you @bigludo7.

@DT-DawidWroblewski could you please review it and if you agree please go ahead and merge it.

**Comment by monamok** [comment]
Dear @DT-DawidWroblewski, would it be possible to get your comments/approval-merge today?
Thanks.

**Comment by DT-DawidWroblewski** [comment]
@monamok working on it...

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [inline code/API_definitions/number_verification.yaml:6]: are we assuming here that MSISDN is somehow returned in Access Token? should it be JWT? where MSISDN is then returned?

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [inline code/API_definitions/number_verification.yaml:9]: In my opinion this API is more End-User related. NaaS is misleading here.

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [inline code/API_definitions/number_verification.yaml:17]: I'd stick to one naming convention, phone number does not equal MSISDN, therefore it can be misleading to audience, e.g. phone number can be fixed number and mobile number. I recommend to use "msisdn" wherever you call phone number or provide a definition of phone number

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml]: I removed it.

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml:16]: I added this to the API description to be more clear that we're refering to the cell phone number not a landline:
_In this API **phone number** term refers to the mobile phone number (MSISDN)_

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml:6]: The API doesn't return the Token. It's a three legged authentication and the phone number is associated to the token so it can be retrieved by the API in order to compare it with the input data.

Comment (DT-DawidWroblewski) [reply code/API_definitions/number_verification.yaml:6]: Hi @monamok ! 

this particular sentence makes confision:

>...verify the number itself by returning the phone number retrieved from the authenticated user's access token.

I read it like:

>as a sw developer i can retrieve phone number from authenticated user's access token, so that I can verify phone number.

maybe sth like this is better:

>as a sw developer I can retrieve information about phone number of the device, so that I can confirm that end user possess the device.

what do yo think?

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [reply code/API_definitions/number_verification.yaml]: thx!

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [reply code/API_definitions/number_verification.yaml:16]: thx!

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml]: I rephrased it. Please take a look :)

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [inline code/API_definitions/number_verification.yaml:15]: phone number retrieved from the access token --> phone number *associated to* the access token

File: code/API_definitions/number_verification.yaml
Comment (DT-DawidWroblewski) [inline code/API_definitions/number_verification.yaml:99]: `device phone number` retrieved from the access token --> `device phone number` *associated to* the access token

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml]: Edited

File: code/API_definitions/number_verification.yaml
Comment (monamok) [reply code/API_definitions/number_verification.yaml]: Edited

**Review by DT-DawidWroblewski — APPROVED** [review]
thx!

**Comment by DT-DawidWroblewski** [comment]
merging under condition that Number Verify based Mobile Connect Verified MSISDN yaml is also available inside PR, although I'm not able to merge it by myself and waiting for @bigludo7 approval.

**Pull request merged**
